corehq/blobs/migrate_metadata.py

```diff
@@ -0,0 +1,323 @@
+        SqlMigrator, BlobMetaMigrator = make_migrators(mod)
+        couch_migrator = partial(BlobMetaMigrator, blob_helper=couch_blob_helper)
+
+        def db_key(doc_type):
+            if isinstance(doc_type, tuple):
+                doc_type = doc_type[1]
+            return doc_type.get_db().dbname
+
+        for key, types in groupby(sorted(self.couch_types, key=db_key), key=db_key):
+            slug = "%s-%s" % (self.slug, key)
+            yield mod.Migrator(slug, list(types), couch_migrator)
+
+        for rex in self.sql_reindexers:
+            slug = "%s-%s" % (self.slug, rex.model_class.__name__)
+            yield SqlMigrator(slug, rex(), BlobMetaMigrator)
+
+    def migrate(self, filename, *args, **kw):
+        def filen(n):
+            return None if filename is None else "{}.{}".format(filename, n)
+        migrated = 0
+        skipped = 0
+        for n, item in enumerate(self.iter_migrators()):
+            one_migrated, one_skipped = item.migrate(filen(n), *args, **kw)
+            migrated += one_migrated
+            skipped += one_skipped
+            print("\n")
+        if not skipped:
+            BlobMigrationState.objects.get_or_create(slug=self.slug)[0].save()
+        return migrated, skipped
+
+
+def make_migrators(mod):
+    # defer class definitions to work around circular import
+
+    class BlobMetaMigrator(mod.BaseDocMigrator):
+        """Migrate blob metadata to BlobMeta model"""
+
+        def __init__(self, *args, **kw):
+            super(BlobMetaMigrator, self).__init__(*args, **kw)
+            self.total_blobs = 0
+
+        def _backup_doc(self, doc):
+            pass
+
+        def _do_migration(self, doc):
+            if not doc.get("external_blobs"):
+                return True
+            type_code = self.get_type_code(doc)
+            obj = self.blob_helper(doc, self.couchdb, type_code)
+            db = get_db_alias_for_partitioned_doc(doc["_id"])
+            with connections[db].cursor() as cursor:
+                for name, meta in six.iteritems(obj.blobs):
+                    if meta.blobmeta_id is not None:
+                        # blobmeta already saved
+                        continue
+                    if obj.domain is None:
+                        print("Unknown domain: {!r}".format(obj))
+                    cursor.execute("""
+                        INSERT INTO blobs_blobmeta (
+                            domain,
+                            type_code,
+                            parent_id,
+                            name,
+                            key,
+                            content_type,
+                            content_length,
+                            created_on
+                        ) VALUES (%s, %s, %s, %s, %s, %s, %s, CLOCK_TIMESTAMP())
+                        ON CONFLICT (key) DO NOTHING
+                    """, params=[
+                        (UNKNOWN_DOMAIN if obj.domain is None else obj.domain),
```

Comment:
will every type have a `domain` attrib? Or should you do `getattr(obj, 'domain', UNKNOWN_DOMAIN)`?

Reply:
looks like they will always have one here

Reply:
Yes, the `domain` attribute should always be present at this point. An assertion error may have been raised earlier if it was not set (for couch docs), but that would indicate a bug.

https://github.com/dimagi/commcare-hq/blob/ed93401415cefdd42b561e6efcf968d7acf58523/corehq/blobs/migrate_metadata.py#L254-L260

For `SqlBlobHelper`

https://github.com/dimagi/commcare-hq/blob/ed93401415cefdd42b561e6efcf968d7acf58523/corehq/blobs/migrate_metadata.py#L156-L158